test/NuGet.Core.Tests/NuGet.Packaging.Test/DependeciesGroupsForEachTFMRuleTests.cs

```diff
@@ -0,0 +1,242 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using NuGet.Common;
+using NuGet.Packaging.Rules;
+using Xunit;
+
+namespace NuGet.Packaging.Test
+{
+    public class DependeciesGroupsForEachTFMRuleTests
```

Comment:
I'm un-resolving this comment because it still stands. Now that `DependeciesGroupsForEachTFMRule.CatagoriseTFMs` and `DependeciesGroupsForEachTFMRule.GenerateWarnings` are independent, I think it would be better (easier!) to test them independently.

One set of tests for `DependeciesGroupsForEachTFMRule.CatagoriseTFMs` that gives the two lists of files and validates that the 3 lists returned contain the expected values. This makes it easier to think of test cases required:
* happy-case, where there are only exact matches
* simple cases where only one of the 3 lists contain TFMs
* complex cases where multiple of the lists contain TFMs.

plus edge/special cases like:
* meta-package (package with no files, only dependencies; ideally should not generate warnings)
* package with no dependencies. I can't remember, did the spec say if we should generate a warning or not?

`DependeciesGroupsForEachTFMRule.GenerateWarnings` should also have its own set of tests. My gut feeling is that there will be fewer test cases to test. Given there's a bunch of code involved in generating the text of the message, the test should not only validate that the correct error code was generated, but the text itself is also correct, so we know all that string concatenation (well, string building) works correctly (check the whole text, not just that the TFM is somewhere in the string, to ensure that all the newline and whitespace is where we expect). Plus the happy case where all 3 lists are empty, no warnings are generated. Since the special cases in `CatagoriseTFMs` generate empty results, which is the same as the happy case, no tests are needed to test `GenerateWarnings` with the special/edge cases.